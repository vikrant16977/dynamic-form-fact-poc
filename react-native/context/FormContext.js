import React, { createContext, useState, useEffect, useRef } from "react";
import axios from "axios";

export const FormContext = createContext();

export const FormProvider = ({ children }) => {
  const [forms, setForms] = useState([]);
  const [selectedFormId, setSelectedFormId] = useState(null);
  const [loading, setLoading] = useState(true);
  const intervalRef = useRef(null);

  const fetchForms = async () => {
    try {
      const response = await axios.get(
        "https://dynamic-form-mobile-noktwd7ra-vikrant-ahers-projects.vercel.app/api/forms",
        {
          headers: {
            "User-Agent": "Mozilla/5.0",
          },
        }
      );
      const data = response.data;
      console.log("Fetched data:", data);

      if (!data || !Array.isArray(data.value)) {
        console.error("Unexpected API response:", data);
        setForms([]);
        setLoading(false);
        return;
      }

      const formsFetched = [];

      for (const item of data.value) {
        try {
          let parsedSchema;

          if (typeof item.schema === "string") {
            let cleanStr = item.schema;

            if (cleanStr.startsWith('"') && cleanStr.endsWith('"')) {
              cleanStr = cleanStr.slice(1, -1);
            }

            cleanStr = cleanStr.replace(/\\"/g, '"');

            parsedSchema = JSON.parse(cleanStr);
          } else {
            parsedSchema = item.schema;
          }

          if (Array.isArray(parsedSchema)) {
            parsedSchema.forEach((form) => {
              formsFetched.push({ ...form, id: item.ID });
            });
          } else {
            formsFetched.push({ ...parsedSchema, id: item.ID });
          }
        } catch (e) {
          console.error("Failed to parse form schema for item:", item);
          console.error("Error:", e);
        }
      }

      setForms(formsFetched);
      if (formsFetched.length > 0) setSelectedFormId(formsFetched[0].id);
    } catch (error) {
      console.error("Failed to fetch forms:", error);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchForms();

    intervalRef.current = setInterval(() => {
      fetchForms();
    }, 240000);

    return () => {
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
      }
    };
  }, []);

  const addNewForm = (title) => {
    const newForm = { id: Date.now(), title, sections: [] };
    setForms((f) => [...f, newForm]);
    setSelectedFormId(newForm.id);
  };

  const addSectionToForm = (formId, section) => {
    setForms((f) =>
      f.map((fm) =>
        fm.id === formId ? { ...fm, sections: [...fm.sections, section] } : fm
      )
    );
  };

  const addQuestionToForm = (formId, sectionId, question) => {
    setForms((f) =>
      f.map((fm) => {
        if (fm.id !== formId) return fm;
        return {
          ...fm,
          sections: fm.sections.map((sec) =>
            sec.id === sectionId
              ? { ...sec, questions: [...sec.questions, question] }
              : sec
          ),
        };
      })
    );
  };

  const updateSelectedFormId = (id) => setSelectedFormId(id);
  const getSelectedForm = () => forms.find((fm) => fm.id === selectedFormId);

  return (
    <FormContext.Provider
      value={{
        forms,
        addNewForm,
        addSectionToForm,
        addQuestionToForm,
        updateSelectedFormId,
        getSelectedForm,
        selectedFormId,
        setForms,
        loading,
      }}
    >
      {children}
    </FormContext.Provider>
  );
};
