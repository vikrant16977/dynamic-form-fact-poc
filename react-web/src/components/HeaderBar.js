import React from "react";
import { Menu, Image, Icon ,Button} from "semantic-ui-react";
import { Link, useLocation } from "react-router-dom";
import logo from "./images/socal-gas.png";

const HeaderBar = () => {
  const location = useLocation();

  return (
    <>
      <Menu borderless style={{ background: "#FFFF", padding: "0 2rem" }}>
        {/* Logo */}
        <Menu.Item as={Link} to="/">
          <Image
            src={logo}
            size="tiny"
            alt="SoCalGas Logo"
            background="#FFFF"
          />
        </Menu.Item>
      </Menu>
      <Menu
        borderless
        fixed="top"
        style={{ background: "#0C2D87", padding: "0 2rem " }}
      >
        {/* Navigation */}
        <Menu.Item
          as={Link}
          to="/"
          name="Home"
          style={{
            color: "#fff",
            fontWeight: location.pathname === "/" ? "bold" : "normal",
          }}
        />
        <Menu.Item
          as={Link}
          to="/admin"
          name="Form Creator"
          style={{
            color: "#fff",
            fontWeight: location.pathname === "/admin" ? "bold" : "normal",
          }}
        />

        {/* Right Side Icons */}
        <Menu.Menu position="right">
          <Menu.Item as={Link} to="/form">
            <Icon name="user icon" style={{ color: "white", marginRight: 4 }} />
            <span style={{ color: "white" }}>User</span>
          </Menu.Item>
          <Menu.Item as={Link} to="/submitted-forms" >
            <Icon name="file icon" style={{ color: "white", marginRight: 4 }} />
            <span style={{ color: "white" }}>View Submitted Forms</span>
          </Menu.Item>   
          <Menu.Item>
            <div
              style={{
                width: "32px",
                height: "32px",
                borderRadius: "50%",
                backgroundColor: "#002366",
                color: "white",
                display: "flex",
                alignItems: "center",
                justifyContent: "center",
                fontWeight: "bold",
              }}
            >
              OG
            </div>
          </Menu.Item>
        </Menu.Menu>
      </Menu>
    </>
  );
};

export default HeaderBar;
