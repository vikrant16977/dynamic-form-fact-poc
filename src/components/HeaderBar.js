import React from "react";
import { Menu, Image, Container } from "semantic-ui-react";
import { Link } from "react-router-dom";

const HeaderBar = () => {
  return (
    <Menu borderless fixed="top" style={{ background: "#002244", color: "#fff" }}>
      <Container>
        <Menu.Item as={Link} to="/">
          <span style={{ fontSize: "1.5em", fontWeight: "bold", color: "#fff" }}>
            Field Audit Collection Tool
          </span>
        </Menu.Item>

        <Menu.Menu position="right">
          <Menu.Item as={Link} to="/admin" style={{ color: "#fff" }}>
            Admin
          </Menu.Item>
          <Menu.Item as={Link} to="/form" style={{ color: "#fff" }}>
            User
          </Menu.Item>
        </Menu.Menu>
      </Container>
    </Menu>
  );
};

export default HeaderBar;
