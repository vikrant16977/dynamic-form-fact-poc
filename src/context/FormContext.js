// src/context/FormContext.js
import React, { createContext, useState, useEffect } from "react";
import { v4 as uuidv4 } from 'uuid';

export const FormContext = createContext();

export const FormProvider = ({ children }) => {
  const [forms, setForms] = useState([]);
  const [activeFormId, setActiveFormId] = useState(null);

  // Load from localStorage
  useEffect(() => {
    const stored = localStorage.getItem("forms");
    if (stored) {
      setForms(JSON.parse(stored));
    }
  }, []);

  // Save to localStorage
  useEffect(() => {
    localStorage.setItem("forms", JSON.stringify(forms));
  }, [forms]);

  const createNewForm = (title) => {
    const newForm = { id: uuidv4(), title, fields: [] };
    setForms([...forms, newForm]);
    setActiveFormId(newForm.id);
  };

  const addFieldToForm = (formId, field) => {
    setForms(forms.map(form =>
      form.id === formId
        ? { ...form, fields: [...form.fields, field] }
        : form
    ));
  };

  const getActiveForm = () => forms.find(form => form.id === activeFormId);

  return (
    <FormContext.Provider value={{
      forms,
      activeFormId,
      createNewForm,
      addFieldToForm,
      setActiveFormId,
      getActiveForm
    }}>
      {children}
    </FormContext.Provider>
  );
};
