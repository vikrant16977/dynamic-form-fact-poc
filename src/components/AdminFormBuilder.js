import React, { useState, useContext } from "react";
import {
  Segment,
  Header,
  Form,
  Button,
  Dropdown,
  Grid,
  Icon
} from "semantic-ui-react";
import { FormContext } from "../context/FormContext";
import FormPreview from "./FormPreview";
import FieldPropertyModal from "./FieldPropertyModal";

const questionTypeOptions = [
  { key: "text", text: "Text Field", value: "text", icon: "edit" },
  { key: "textarea", text: "Textarea", value: "textarea", icon: "align left" },
  { key: "radio", text: "Radio Buttons", value: "radio", icon: "dot circle" },
  { key: "checkbox", text: "Checkbox", value: "checkbox", icon: "check square" },
  { key: "dropdown", text: "Dropdown", value: "dropdown", icon: "dropdown" },
  { key: "number", text: "Number Field", value: "number", icon: "hashtag" },
  { key: "email", text: "Email Field", value: "email", icon: "envelope" },
  { key: "date", text: "Date Picker", value: "date", icon: "calendar alternate" },
];

const AdminFormBuilder = ({
  selectedForm,
  forms,
  setForms,
  selectedFormId,
}) => {
  const [newSectionTitle, setNewSectionTitle] = useState("");
  const [activeSectionId, setActiveSectionId] = useState(null);
  const [modalOpen, setModalOpen] = useState(false);
  const [selectedField, setSelectedField] = useState(null);
  const [selectedSectionId, setSelectedSectionId] = useState(null);
  const [sectionCounter, setSectionCounter] = useState(1);
  const [questionCounter, setQuestionCounter] = useState(1);

  const { addSectionToForm } = useContext(FormContext);

  const handleQuestionClick = (sectionId, questionId) => {
    const section = selectedForm.sections.find((s) => s.id === sectionId);
    const question = section.questions.find((q) => q.id === questionId);
    setSelectedField(question);
    setSelectedSectionId(sectionId);
    setModalOpen(true);
  };

  const handleModalSave = (updatedField) => {
    const updatedForms = forms.map((form) => {
      if (form.id === selectedFormId) {
        const updatedSections = form.sections.map((section) => {
          if (section.id === selectedSectionId) {
            const updatedQuestions = section.questions.map((q) =>
              q.id === updatedField.id ? updatedField : q
            );
            return { ...section, questions: updatedQuestions };
          }
          return section;
        });
        return { ...form, sections: updatedSections };
      }
      return form;
    });

    setForms(updatedForms);
    localStorage.setItem("forms", JSON.stringify(updatedForms));
    setModalOpen(false);
  };

  const handleAddSection = () => {
    if (newSectionTitle.trim() !== "") {
      const newSection = {
        id: `sec${sectionCounter}`,
        sectionTitle: newSectionTitle.trim(),
        questions: [],
      };
      addSectionToForm(selectedFormId, newSection);
      setNewSectionTitle("");
      setSectionCounter(sectionCounter + 1);
    }
  };

  const handleAddFieldToSection = (type) => {
    if (!activeSectionId) {
      alert("Please select a section to add the field.");
      return;
    }

    const newQuestion = {
      id: `q${questionCounter}`,
      label: `${type} field`,
      type,
      options:
        type === "radio" || type === "dropdown"
          ? ["Option 1", "Option 2"]
          : [],
      required: true,
    };
    setQuestionCounter(questionCounter + 1);

    const updatedForms = forms.map((form) => {
      if (form.id === selectedFormId) {
        const updatedSections = form.sections.map((section) => {
          if (section.id === activeSectionId) {
            return {
              ...section,
              questions: [...section.questions, newQuestion],
            };
          }
          return section;
        });
        return { ...form, sections: updatedSections };
      }
      return form;
    });

    setForms(updatedForms);
    localStorage.setItem("forms", JSON.stringify(updatedForms));
  };

  return (
    <>
      <Grid>
        <Grid.Column width={"3"}>
          <Segment>
            <Header as="h4">Add Section</Header>
            <Form>
              <Form.Input
                placeholder="Section Title"
                value={newSectionTitle}
                onChange={(e) => setNewSectionTitle(e.target.value)}
              />
              <Button color="blue" onClick={handleAddSection} fluid>
                Add Section
              </Button>
            </Form>
          </Segment>
          <Segment>
            <Header as="h4">Add Components</Header>
            {questionTypeOptions.map((opt) => (
              <Button
                key={opt.key}
                icon
                labelPosition="left"
                onClick={() => handleAddFieldToSection(opt.value)}
                style={{ margin: "0.25rem 0" }}
                fluid
              >
                <Icon name={opt.icon} />
                {opt.text}
              </Button>
            ))}
          </Segment>
        </Grid.Column>

        <Grid.Column width={"13"}>
          <Header as="h3" style={{ textDecoration: "underline" }}>
            Live Preview
          </Header>
          {selectedForm?.sections?.length > 0 ? (
            <Header
              as="h3"
              size="tiny"
              color="grey"
              style={{ marginTop: "-0.5rem" }}
            >
              Click on the below field types to open its editing options*
            </Header>
          ) : (
            <Header
              as="h3"
              size="tiny"
              color="grey"
              style={{ marginTop: "-0.5rem" }}
            >
              Add sections from left menu by entering section's title and
              clicking on "Add Section" button to start creating your form.
            </Header>
          )}
          <FormPreview
            form={selectedForm}
            setActiveSectionId={setActiveSectionId}
            activeSectionId={activeSectionId}
            onQuestionClick={handleQuestionClick}
          />
        </Grid.Column>
      </Grid>
      <FieldPropertyModal
        open={modalOpen}
        onClose={() => setModalOpen(false)}
        onSave={handleModalSave}
        field={selectedField}
      />
    </>
  );
};

export default AdminFormBuilder;
