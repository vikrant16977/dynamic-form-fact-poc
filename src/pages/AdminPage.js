// src/pages/AdminPage.js
import React, { useContext, useState } from 'react';
import { Container, Header, Divider, List, Button, Input, Segment } from 'semantic-ui-react';
import { FormContext } from '../context/FormContext';
import FormBuilder from '../components/FormBuilder';
import { Link } from 'react-router-dom';

const AdminPage = () => {
  const {
    forms,
    createNewForm,
    activeFormId,
    setActiveFormId,
    getActiveForm
  } = useContext(FormContext);

  const [newFormTitle, setNewFormTitle] = useState("");

  const handleCreateForm = () => {
    if (newFormTitle.trim()) {
      createNewForm(newFormTitle);
      setNewFormTitle("");
    }
  };

  const activeForm = getActiveForm();

  return (
    <Container style={{ marginTop: '2em' }}>
      <Header as="h2">Admin - Form Manager</Header>

      <Segment>
        <Input
          placeholder="New Form Title"
          value={newFormTitle}
          onChange={(e) => setNewFormTitle(e.target.value)}
        />
        <Button onClick={handleCreateForm} color="green" style={{ marginLeft: '1em' }}>
          Create Form
        </Button>
      </Segment>

      <Header as="h4">Your Forms:</Header>
      <List divided relaxed>
        {forms.map((form) => (
          <List.Item key={form.id}>
            <List.Content floated="right">
              <Button
                size="mini"
                onClick={() => setActiveFormId(form.id)}
                color={form.id === activeFormId ? "blue" : "grey"}
              >
                Edit
              </Button>
            </List.Content>
            <List.Content>
              <List.Header>{form.title}</List.Header>
              <List.Description>{form.fields.length} field(s)</List.Description>
            </List.Content>
          </List.Item>
        ))}
      </List>

      <Divider />

      {activeForm ? (
        <>
          <Header as="h3">Editing: {activeForm.title}</Header>
          <FormBuilder />
        </>
      ) : (
        <p>Select a form to edit its fields</p>
      )}

      <Divider />
      <Button as={Link} to="/form" color="teal">
        Go to User Form â†’
      </Button>
    </Container>
  );
};

export default AdminPage;
