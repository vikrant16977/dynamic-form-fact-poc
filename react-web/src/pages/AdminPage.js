// src/pages/AdminPage.js
import React, { useContext, useState } from "react";
import { FormContext } from "../context/FormContext";
import {
  Container,
  Header,
  Form,
  Button,
  Dropdown,
  Segment,
  Grid,
  Divider,
  Icon,
} from "semantic-ui-react";
import { Link, useNavigate } from "react-router-dom";
import HeaderBar from "../components/HeaderBar";
import AdminFormBuilder from "../components/AdminFormBuilder";

const AdminPage = () => {
  const navigate = useNavigate();
  const { forms, setForms, selectedFormId, addNewForm, updateSelectedFormId } =
    useContext(FormContext);
  const [newFormTitle, setNewFormTitle] = useState("");
  const [isSubmitting, setIsSubmitting] = useState(false);
  const selectedForm = forms.find((f) => f.id === selectedFormId);
  const newForm = localStorage.getItem("newForm");

  const handleCreateForm = () => {
    if (!newFormTitle.trim()) return;
    addNewForm(newFormTitle.trim());
    setNewFormTitle("");
  };

const generateShortId = (prefix) => {
  const random = Math.random().toString(36).substring(2, 4); // 2 characters
  return `${prefix}${random}`; // e.g., fz9
};

  const handleConfirmForm = async () => {
    if (!selectedForm) {
      alert("No form selected to submit.");
      return;
    }

    const payload = {
     
      ID:Math.floor(1000 + Math.random() * 9000), 
      title: selectedForm.title,
      description: selectedForm.description || "Form created via builder",
      schema: JSON.stringify(selectedForm),
    };

    setIsSubmitting(true);
    try {
      const response = await fetch("https://dynamicformbackend.onrender.com/odata/v4/catalog/Forms", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      if (response.ok) {
        const result = await response.json();
        console.log("Form submitted successfully:", result);
        alert("Form submitted successfully!");
      } else {
        console.error("Failed to submit form:", response.statusText);
        alert("Failed to submit form.");
      }
    } catch (error) {
      console.error("Error submitting form:", error);
      alert("Error submitting form.");
    } finally {
      setIsSubmitting(false);
    }
  };

  return (
    <>
      <HeaderBar />
      <Container style={{ marginTop: "6rem" }}>
        <Segment>
          {newForm === "true" ? (
            <Grid columns={2} verticalAlign="middle">
              <Grid.Column>
                <Header as="h2">Form Builder</Header>
              </Grid.Column>

              <Grid.Column textAlign="right">
                <Button
                  color="grey"
                  style={{ marginRight: "1rem" }}
                  onClick={() => navigate("/home")}
                >
                  Back
                </Button>
                <Button
                  color="blue"
                  style={{ marginRight: "1rem" }}
                  onClick={handleCreateForm}
                >
                  Create
                </Button>
                <Button
                  color="green"
                  onClick={handleConfirmForm}
                  loading={isSubmitting}
                  disabled={isSubmitting}
                >
                  <Icon name="check" />
                  Confirm & Submit
                </Button>
              </Grid.Column>
            </Grid>
          ) : (
            <Grid columns={2} verticalAlign="middle">
              <Grid.Column>
                <Header as="h2">Form Editor</Header>
              </Grid.Column>
            </Grid>
          )}
          <Divider />

          {newForm === "true" && (
            <Form>
              <Form.Input
                placeholder="Form Title"
                value={newFormTitle}
                onChange={(e) => setNewFormTitle(e.target.value)}
              />
            </Form>
          )}

          {forms.length > 0 && (
            <Form style={{ marginTop: "1rem" }}>
              <Form.Field>
                <label>Select Form to Edit</label>
                <Dropdown
                  placeholder="Select a form"
                  fluid
                  selection
                  options={forms.map((f) => ({
                    key: f.id,
                    text: f.title,
                    value: f.id,
                  }))}
                  value={selectedFormId}
                  onChange={(_, { value }) => updateSelectedFormId(value)}
                />
              </Form.Field>
            </Form>
          )}
        </Segment>

        {selectedFormId && selectedForm && (
          <Grid>
            <Grid.Column width={"16"}>
              <AdminFormBuilder
                selectedForm={selectedForm}
                forms={forms}
                setForms={setForms}
                selectedFormId={selectedFormId}
              />
            </Grid.Column>
          </Grid>
        )}
      </Container>
    </>
  );
};

export default AdminPage;
