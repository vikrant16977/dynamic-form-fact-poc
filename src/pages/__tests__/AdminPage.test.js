import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import AdminPage from "../AdminPage";
import { FormContext } from "../../context/FormContext";
import { MemoryRouter } from "react-router-dom";
import HeaderBar from "../../components/HeaderBar";
import AdminFormBuilder from "../../components/AdminFormBuilder";

// Mock dependencies
jest.mock("../../components/HeaderBar", () => () => (
  <div data-testid="header-bar" />
));
jest.mock("../../components/AdminFormBuilder", () => (props) => (
  <div data-testid="admin-form-builder">{JSON.stringify(props)}</div>
));
const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
  Link: ({ to, children }) => <a href={to}>{children}</a>,
}));

const renderAdminPage = ({
  forms = [],
  setForms = jest.fn(),
  selectedFormId = null,
  addNewForm = jest.fn(),
  updateSelectedFormId = jest.fn(),
  localStorageNewForm = null,
} = {}) => {
  if (localStorageNewForm !== null) {
    window.localStorage.setItem("newForm", localStorageNewForm);
  } else {
    window.localStorage.removeItem("newForm");
  }
  return render(
    <FormContext.Provider
      value={{
        forms,
        setForms,
        selectedFormId,
        addNewForm,
        updateSelectedFormId,
      }}
    >
      <MemoryRouter>
        <AdminPage />
      </MemoryRouter>
    </FormContext.Provider>
  );
};

describe("AdminPage", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    window.localStorage.clear();
  });

  it("renders HeaderBar and main containers", () => {
    renderAdminPage();
    expect(screen.getByTestId("header-bar")).toBeInTheDocument();
    expect(screen.getByText(/Form Editor/i)).toBeInTheDocument();
  });

  it("displays Form Builder section when newForm is true", () => {
    renderAdminPage({ localStorageNewForm: "true" });
    expect(screen.getByText(/Form Builder/i)).toBeInTheDocument();
    expect(screen.getByText(/Back/i)).toBeInTheDocument();
    expect(screen.getByText(/Create/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/Form Title/i)).toBeInTheDocument();
  });

  it("calls navigate when Back button is clicked", () => {
    renderAdminPage({ localStorageNewForm: "true" });
    fireEvent.click(screen.getByText(/Back/i));
    expect(mockNavigate).toHaveBeenCalledWith("/home");
  });

  it("handles input and Create button, calls addNewForm, then clears input", () => {
    const addNewForm = jest.fn();
    renderAdminPage({
      localStorageNewForm: "true",
      addNewForm,
    });
    const input = screen.getByPlaceholderText(/Form Title/i);
    fireEvent.change(input, { target: { value: "Test Form" } });
    expect(input.value).toBe("Test Form");
    fireEvent.click(screen.getByText(/Create/i));
    expect(addNewForm).toHaveBeenCalledWith("Test Form");
    expect(input.value).toBe(""); // Should be cleared after create
  });

  it("does not call addNewForm if input is empty or whitespace", () => {
    const addNewForm = jest.fn();
    renderAdminPage({
      localStorageNewForm: "true",
      addNewForm,
    });
    const input = screen.getByPlaceholderText(/Form Title/i);
    fireEvent.change(input, { target: { value: "   " } });
    fireEvent.click(screen.getByText(/Create/i));
    expect(addNewForm).not.toHaveBeenCalled();
  });

  it("shows form dropdown when there is at least one form", () => {
    const forms = [
      { id: 1, title: "Form 1" },
      { id: 2, title: "Form 2" },
    ];
    renderAdminPage({ forms });
    expect(screen.getByText(/Select Form to Edit/i)).toBeInTheDocument();
    expect(screen.getByText("Form 1")).toBeInTheDocument();
    expect(screen.getByText("Form 2")).toBeInTheDocument();
  });

  it("calls updateSelectedFormId when dropdown value changes", () => {
    const forms = [{ id: 1, title: "Form 1" }];
    const updateSelectedFormId = jest.fn();
    renderAdminPage({
      forms,
      selectedFormId: null,
      updateSelectedFormId,
    });
    const dropdown = screen.getByRole("listbox");
    fireEvent.click(dropdown);
    const option = screen.getByText("Form 1");
    fireEvent.click(option);
    expect(updateSelectedFormId).toHaveBeenCalledWith(1);
  });

  it("renders AdminFormBuilder when selectedFormId matches a form", () => {
    const forms = [{ id: 5, title: "Form X" }];
    renderAdminPage({
      forms,
      selectedFormId: 5,
    });
    expect(screen.getByTestId("admin-form-builder")).toBeInTheDocument();
  });

  it("does not render AdminFormBuilder when no form is selected", () => {
    renderAdminPage({
      forms: [{ id: 1, title: "Form 1" }],
      selectedFormId: null,
    });
    expect(screen.queryByTestId("admin-form-builder")).toBeNull();
  });

  it("renders correctly for empty forms state (edge case)", () => {
    renderAdminPage({ forms: [] });
    expect(screen.queryByText(/Select Form to Edit/i)).toBeNull();
  });

  it("handles edge case: selectedFormId doesn't match any form", () => {
    renderAdminPage({
      forms: [{ id: 1, title: "Form 1" }],
      selectedFormId: 999,
    });
    expect(screen.queryByTestId("admin-form-builder")).toBeNull();
  });

  it("renders loading state if context is slow (simulate)", async () => {
    // Simulate context loading by not providing forms
    render(
      <FormContext.Provider value={{ forms: [] }}>
        <MemoryRouter>
          <AdminPage />
        </MemoryRouter>
      </FormContext.Provider>
    );
    // Should not crash; may not render anything meaningful
    expect(screen.getByTestId("header-bar")).toBeInTheDocument();
  });

  it("snapshot test: initial render, no newForm, no forms", () => {
    const { asFragment } = renderAdminPage();
    expect(asFragment()).toMatchSnapshot();
  });
});
