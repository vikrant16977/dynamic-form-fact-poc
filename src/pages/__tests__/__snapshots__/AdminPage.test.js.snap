// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPage snapshot test: initial render, no newForm, no forms 1`] = `
<DocumentFragment>
  <div
    data-testid="header-bar"
  />
  <div
    class="ui container"
    style="margin-top: 6rem;"
  >
    <div
      class="ui segment"
    >
      <div
        class="ui middle aligned two column grid"
      >
        <div
          class="column"
        >
          <h2
            class="ui header"
          >
            Form Editor
          </h2>
        </div>
         
      </div>
      <div
        class="ui divider"
      />
    </div>
  </div>
</DocumentFragment>
`;
